### PR TITLE
ui: don't show exception for "Too many files" error

### PR DESCRIPTION
**Pull request by efiop**
https://github.com/iterative/dvc.org/pull/875#discussion_r362920851

Stops showing the exception in non-verbose mode, but shows full
traceback in verbose mode.

Introducing our custom parameter `tb_only` through standard `extra`
logging parameter.

* [x] ❗ Have you followed the guidelines in the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) list?

* [x] 📖 Check this box if this PR **does not** require [documentation](https://dvc.org/doc) updates, or if it does **and** you have created a separate PR in [dvc.org](https://github.com/iterative/dvc.org) with such updates (or at least opened an issue about it in that repo). Please link below to your PR (or issue) in the [dvc.org](https://github.com/iterative/dvc.org) repo.

* [x] ❌ Have you checked DeepSource, CodeClimate, and other sanity checks below? We consider their findings recommendatory and don't expect everything to be addressed. Please review them carefully and fix those that actually improve code or fix bugs.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

### CODE REVIEW

**Comment by shcheklein** [comment]
@efiop can include the output after this change, please?

**Comment by efiop** [comment]
@shcheklein not sure I follow, could you rephrase, please?

File: dvc/logger.py [hunk 1]
```diff
@@ -62,7 +63,7 @@ def format(self, record):
             return record.msg
 
         if record.levelname == "ERROR" or record.levelname == "CRITICAL":
-            exception, stack_trace = self._parse_exc(record.exc_info)
+            exception, stack_trace = self._parse_exc(record)
```
Comment (efiop) [inline dvc/logger.py]: Also looked into standard `formatException` again (since py2 was dropped), but it doesn't quite do what we need. :slightly_frowning_face:

**Comment by shcheklein** [comment]
> @shcheklein not sure I follow, could you rephrase, please?

I mean the CLI output after this changes vs what was before. The same way as @pared included then in the docs discussion. It's hard otherwise to see the difference and understand how does it looks like.

**Comment by efiop** [comment]
@shcheklein Ah, right. So before:
```
ERROR: too many open files, please increase your `ulimit` - [Errno 24] Too many open files: '/home/prd/development/projects/playground/2473/repo/.dvc/cache/10/788f0c3b6f08ef6
f183f49490183d0'


Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
```
after:
```
ERROR: too many open files, please increase your `ulimit`


Having any troubles? Hit us up at https://dvc.org/support, we are always happy to help!
```

**Comment by shcheklein** [comment]
@efiop so, i'm lost then :) how does it correspond with this one - https://github.com/iterative/dvc.org/pull/875#discussion_r362733001 ? was it just a suggestion?

**Comment by efiop** [comment]
@shcheklein I've removed the unnecessary second `Too many open files` and the filename. I didn't adjust the message itself, as it will be handled by @pared in #3017

**Comment by shcheklein** [comment]
@efiop got it! thanks.

File: dvc/logger.py (hunk 1 above)
Comment (mroutis) [reply dvc/logger.py]: yeah :disappointed:  I was working on a PR to refactor that but couldn't make it work as intended.

**Comment by mroutis** [comment]
@efiop , what about showing the whole stack every time?

**Comment by efiop** [comment]
@mroutis what about it? We don't unless `-v` is specified, the same as it was before.

**Comment by mroutis** [comment]
Why are you passing the `stack_info=True` if you are not showing it on https://github.com/iterative/dvc/pull/3058#issuecomment-570854778 ?

**Comment by efiop** [comment]
@mroutis I do, but only on `-v`, because it is our policy. I should've explicitly clarified, that default logger.exception/error methods have `exc_info` and `stack_info` params. For `error` exc_info is None by default, so it doesn't show it and `stack_info` is False by default for both error/exception. So my change takes advantage of those defaults so that `error` get's `exc_info=None`, but `stack_info=True`, which tells it to not show the exception but show its traceback.

**Comment by mroutis** [comment]
@efiop , it is hard to me to review this :thinking: I have mixed feelings about the way we are handling exceptions and displaying error messages.

I think the `don't show exception for "Too many files" error could be solved by doing `raise ... from None` to delete the cause of the exception.

`[Errno 24] Too many open files: '/home/prd/development/projects/playground/2473/repo/.dvc/cache/10/788f0c3b6f08ef6f183f49490183d0'` is the exception's `.__cause__`.

The `"Too many files" is the message that we are displaying with `logger.exception("")`.

Still not sure why we need to deal with the stack trace?

**Comment by efiop** [comment]
> I think the don't show exception for "Too many files" error could be solved by doing raise ... from None` to delete the cause of the exception.

But raise where? We don't have control over it, so we will have to hack to delete the exception.

Also, that would remove the traceback from `-v` output too, which is harmful, as it will deprive us of debug info.

**Comment by mroutis** [comment]
> Also, that would remove the traceback from -v output too, which is harmful, as it will deprive us of debug info.

True that.

Anyways, using `stack_info` seems like a hack, since we are not displaying it.

I'd better opt to add an extra argument to the exception itself and handle it correctly on the logger.

```python
    except Exception as exc:  # pylint: disable=broad-except
        if isinstance(exc, OSError) and exc.errno == errno.EMFILE:
            exc.display_only_message = True

            logger.exception(
                "too many open files, please increase your `ulimit`"
            )
```

**Comment by efiop** [comment]
> Anyways, using stack_info seems like a hack, since we are not displaying it.

We NEVER display tracebacks in non-verbose mode, hence why we have all that code in logger.py to handle it. `stack_info` is a standard mechanism to tell logger to show the stacktrace and it is part of the standard to use it with `exc_info` in any combination.  (see https://docs.python.org/3/library/logging.html#logging.Logger.debug) Our current formatter doesn't process `stack_info` argument at all(it does use `exc_info` though), so in this PR I'm using it to achieve a behaviour that is fairly close to the default behaviour but with our twist on it. If the naming is an issue, we could totally use `extra` dictionary instead (also a default thing meant for custom params) and have `{"tb_only": False}` in it or something like that.

> I'd better opt to add an extra argument to the exception itself and handle it correctly on the logger.

Now that's a hack :slightly_smiling_face:

**Comment by mroutis** [comment]
> We NEVER display tracebacks in non-verbose mode, hence why we have all that code in logger.py to handle it. stack_info is a standard mechanism to tell logger to show the stacktrace and it is part of the standard to use it with exc_info in any combination. (see https://docs.python.org/3/library/logging.html#logging.Logger.debug) Our current formatter doesn't process stack_info argument at all(it does use exc_info though), so in this PR I'm using it to achieve a behaviour that is fairly close to the default behaviour but with our twist on it.

@efiop , agree with your comments, your patch solve the problem :slightly_smiling_face: 

I just felt that it is misleading that we were using `stack_info=True` but not showing the stack info on `--verbose`

If we were doing so, the message would be something like: `Stack (most recent call last):` and the frames would be different.

> If the naming is an issue, we could totally use extra dictionary instead (also a default thing meant for custom params) and have {"tb_only": False} in it or something like that.

Isn't  `extra` dict just for formatting? how could you access it?

**Comment by efiop** [comment]
@mroutis `extra` params just become attributes in the log record. You are right about `stack_info` being cryptic, so I've adjusted my PR to use `tb_only` instead, which does look much nicer to me. Please take a look. :slightly_smiling_face:
